integration-tests/src/test/java/com/datastax/oss/driver/core/TCPFlowControlIT.java

```diff
@@ -0,0 +1,251 @@
+      for (int i = 0; i < NUMBER_OF_SUBMITTED_REQUESTS; i++) {
+        pendingRequests.add(session.executeAsync(SimpleStatement.newInstance(queryString)));
+      }
+
+      // Non-paused nodes should process the requests correctly
+      for (Node n : nonPausedNodes) {
+        Awaitility.await()
+            .until(
+                () -> {
+                  int inFlightRequests = getInFlightRequests(session, n);
+                  return inFlightRequests == 0;
+                });
+      }
+
+      // There are still requests that haven't been written
+      waitForWriteQueueToStabilize(session);
+      assertThat(getWriteQueueSize(session)).isGreaterThan(1);
+
+      SIMULACRON_RULE.cluster().resumeRead();
+
+      int numberOfFinished = 0;
+      for (CompletionStage<AsyncResultSet> result : pendingRequests) {
+        result.toCompletableFuture().get(1, TimeUnit.SECONDS);
+        numberOfFinished = numberOfFinished + 1;
+      }
+
+      int writeQueueSize = getWriteQueueSize(session);
+      assertThat(writeQueueSize).isEqualTo(0);
+      assertThat(numberOfFinished).isEqualTo(NUMBER_OF_SUBMITTED_REQUESTS);
+    }
+  }
+
+  @Test
+  public void should_timeout_requests_when_the_server_paused_reading_without_resuming()
+      throws InterruptedException {
+    byte[] buffer = new byte[10240];
+    Arrays.fill(buffer, (byte) 1);
+    ByteBuffer buffer10Kb = ByteBuffer.wrap(buffer);
+
+    String queryString =
+        String.format("INSERT INTO table1 (id) VALUES (%s)", ByteUtils.toHexString(buffer10Kb));
+    Query query = new Query(queryString, emptyList(), emptyMap(), emptyMap());
+
+    SIMULACRON_RULE.cluster().prime(when(query).then(noRows()));
+
+    try (CqlSession session = SessionUtils.newSession(SIMULACRON_RULE)) {
+      SIMULACRON_RULE.cluster().pauseRead();
+
+      List<CompletionStage<AsyncResultSet>> pendingRequests = new ArrayList<>();
+      for (int i = 0; i < NUMBER_OF_SUBMITTED_REQUESTS; i++) {
+        pendingRequests.add(session.executeAsync(SimpleStatement.newInstance(queryString)));
+      }
+
+      // Assert that there are still requests that haven't been written
+      waitForWriteQueueToStabilize(session);
+      assertThat(getWriteQueueSize(session)).isGreaterThan(1);
+
+      // do not resumeRead
+      for (CompletionStage<AsyncResultSet> result : pendingRequests) {
+        assertThatThrownBy(() -> result.toCompletableFuture().get())
+            .isInstanceOf(ExecutionException.class)
+            .hasCauseInstanceOf(DriverTimeoutException.class);
+      }
+
+      // write queue size is still non empty
+      int writeQueueSize = getWriteQueueSize(session);
+      assertThat(writeQueueSize).isGreaterThan(1);
+      SIMULACRON_RULE.cluster().stop();
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private int getInFlightRequests(CqlSession session, Node n) {
+    return ((Gauge<Integer>)
+            session.getMetrics().get().getNodeMetric(n, DefaultNodeMetric.IN_FLIGHT).get())
+        .getValue();
```

Comment:
You can pass an explicit type argument to `getNodeMetric` call to avoid the unchecked cast. Also using the optional API:
```suggestion
    return session
        .getMetrics()
        .flatMap(metrics -> metrics.<Gauge<Integer>>getNodeMetric(n, DefaultNodeMetric.IN_FLIGHT))
        .map(Gauge::getValue)
        .orElseThrow(() -> new AssertionError("Expected in-flight metric to be enabled"));
```